File: 1/2021.04.07/316.Remove Duplicate Letters/python/solution.py
class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        stack = []
        count = [0] * 26
        in_stack = [False] * 26
        for c in s:
            count[ord(c) - ord('a')] += 1
        for c in s:
            count[ord(c) - ord('a')] -= 1
            if in_stack[ord(c) - ord('a')]:
                continue
            while stack and stack[-1] > c and count[ord(stack[-1]) - ord('a')] > 0:
                in_stack[ord(stack.pop()) - ord('a')] = False
            stack.append(c)
            in_stack[ord(c) - ord('a')] = True
        return ''.join(stack)

File: 1/2021.04.07/316.Remove Duplicate Letters/java/solution.java
class Solution {
    public String removeDuplicateLetters(String s) {
        Deque<Character> stack = new ArrayDeque<>();
        boolean[] inStack = new boolean[26];
        int[] count = new int[26];
        for (char c : s.toCharArray()) {
            count[c - 'a']++;
        }
        for (char c : s.toCharArray()) {
            count[c - 'a']--;
            if (inStack[c - 'a']) {
                continue;
            }
            while (!stack.isEmpty() && stack.peek() > c && count[stack.peek() - 'a'] > 0) {
                inStack[stack.pop() - 'a'] = false;
            }
            stack.push(c);
            inStack[c - 'a'] = true;
        }
        StringBuilder sb = new StringBuilder();
        while (!stack.isEmpty()) {
            sb.append(stack.pop());
        }
        return sb.reverse().toString();
    }
}
